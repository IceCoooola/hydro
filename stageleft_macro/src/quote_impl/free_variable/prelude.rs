use std::collections::HashSet;

use lazy_static::lazy_static;

lazy_static! {
    static ref PRELUDE: HashSet<&'static str> = {
        vec!["Some", "None", "Ok", "Err"]
            .into_iter()
            .collect::<HashSet<&'static str>>()
    };
}

pub fn is_prelude(ident: &syn::Ident) -> bool {
    let ident_str = ident.to_string();
    PRELUDE.contains(&ident_str.as_str())
}
