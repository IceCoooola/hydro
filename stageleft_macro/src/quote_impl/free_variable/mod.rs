use std::collections::{BTreeSet, HashSet};

mod prelude;
use prelude::is_prelude;

#[derive(Debug)]
pub struct ScopeStack {
    scopes: Vec<(HashSet<String>, HashSet<String>)>,
}

impl Default for ScopeStack {
    fn default() -> Self {
        ScopeStack {
            scopes: vec![(HashSet::new(), HashSet::new())],
        }
    }
}

impl ScopeStack {
    pub fn push(&mut self) {
        self.scopes.push((HashSet::new(), HashSet::new()));
    }

    pub fn pop(&mut self) {
        self.scopes.pop();
    }

    pub fn insert_term(&mut self, ident: syn::Ident) {
        self.scopes
            .last_mut()
            .expect("Scope stack should not be empty")
            .0
            .insert(ident.to_string());
    }

    pub fn insert_type(&mut self, ident: syn::Ident) {
        self.scopes
            .last_mut()
            .expect("Scope stack should not be empty")
            .1
            .insert(ident.to_string());
    }

    pub fn contains_term(&self, ident: &syn::Ident) -> bool {
        let ident = ident.to_string();
        self.scopes
            .iter()
            .rev()
            .any(|scope| scope.0.contains(&ident))
    }

    pub fn contains_type(&self, ident: &syn::Ident) -> bool {
        let ident = ident.to_string();
        self.scopes
            .iter()
            .rev()
            .any(|scope| scope.1.contains(&ident))
    }
}

#[derive(Default)]
pub struct FreeVariableVisitor {
    pub free_variables: BTreeSet<syn::Ident>,
    pub current_scope: ScopeStack,
}

impl<'ast> syn::visit::Visit<'ast> for FreeVariableVisitor {
    fn visit_expr_closure(&mut self, i: &'ast syn::ExprClosure) {
        self.current_scope.push();
        i.inputs.iter().for_each(|input| {
            self.visit_pat(input);
        });

        syn::visit::visit_expr_closure(self, i);

        self.current_scope.pop();
    }

    fn visit_item_fn(&mut self, i: &'ast syn::ItemFn) {
        self.current_scope.push();
        syn::visit::visit_item_fn(self, i);
        self.current_scope.pop();
    }

    fn visit_generic_param(&mut self, i: &'ast syn::GenericParam) {
        match i {
            syn::GenericParam::Type(type_param) => {
                self.current_scope.insert_type(type_param.ident.clone());
            }
            syn::GenericParam::Lifetime(lifetime_param) => {
                self.current_scope
                    .insert_type(lifetime_param.lifetime.ident.clone());
            }
            syn::GenericParam::Const(const_param) => {
                self.current_scope.insert_type(const_param.ident.clone());
            }
        }
    }

    fn visit_block(&mut self, i: &'ast syn::Block) {
        self.current_scope.push();
        syn::visit::visit_block(self, i);
        self.current_scope.pop();
    }

    fn visit_local(&mut self, i: &'ast syn::Local) {
        i.init.iter().for_each(|init| {
            syn::visit::visit_local_init(self, init);
        });

        match &i.pat {
            syn::Pat::Ident(pat_ident) => {
                self.current_scope.insert_term(pat_ident.ident.clone());
            }
            syn::Pat::Type(pat_type) => {
                self.visit_pat(&pat_type.pat);
            }
            syn::Pat::Wild(_) => {
                // Do nothing
            }
            _ => panic!("Local variables must be identifiers, got {:?}", i.pat),
        }
    }

    fn visit_ident(&mut self, i: &'ast proc_macro2::Ident) {
        if !self.current_scope.contains_term(i) {
            self.free_variables.insert(i.clone());
        }
    }

    fn visit_lifetime(&mut self, i: &'ast syn::Lifetime) {
        if !self.current_scope.contains_type(&i.ident) {
            self.free_variables.insert(i.ident.clone());
        }
    }

    fn visit_path(&mut self, i: &'ast syn::Path) {
        if i.leading_colon.is_none() && !is_prelude(&i.segments.first().unwrap().ident) {
            let node = i.segments.first().unwrap();
            if i.segments.len() == 1 && !self.current_scope.contains_term(&node.ident) {
                self.free_variables.insert(node.ident.clone());
            }
        }

        for node in i.segments.iter() {
            self.visit_path_arguments(&node.arguments);
        }
    }

    fn visit_arm(&mut self, i: &'ast syn::Arm) {
        self.current_scope.push();
        syn::visit::visit_arm(self, i);
        self.current_scope.pop();
    }

    fn visit_field_pat(&mut self, i: &'ast syn::FieldPat) {
        for it in &i.attrs {
            self.visit_attribute(it);
        }
        self.visit_pat(&i.pat);
    }

    fn visit_pat_ident(&mut self, i: &'ast syn::PatIdent) {
        self.current_scope.insert_term(i.ident.clone());
    }

    fn visit_expr_method_call(&mut self, i: &'ast syn::ExprMethodCall) {
        syn::visit::visit_expr(self, &i.receiver);
        for arg in &i.args {
            self.visit_expr(arg);
        }
    }

    fn visit_type(&mut self, _: &'ast syn::Type) {}

    fn visit_expr_struct(&mut self, node: &'ast syn::ExprStruct) {
        for it in &node.attrs {
            self.visit_attribute(it);
        }
        if let Some(it) = &node.qself {
            self.visit_qself(it);
        }
        // No need to capture the struct path
        // self.visit_path(&node.path);
        for el in syn::punctuated::Punctuated::pairs(&node.fields) {
            let it = el.value();
            self.visit_expr(&it.expr);
        }
        if let Some(it) = &node.rest {
            self.visit_expr(it);
        }
    }

    fn visit_expr_field(&mut self, i: &'ast syn::ExprField) {
        self.visit_expr(&i.base);
    }

    fn visit_macro(&mut self, i: &'ast syn::Macro) {
        // TODO(shadaj): emit a warning if our guess at parsing fails
        match i.delimiter {
            syn::MacroDelimiter::Paren(_binding_0) => i
                .parse_body_with(
                    syn::punctuated::Punctuated::<syn::Expr, syn::Token![,]>::parse_terminated,
                )
                .ok()
                .iter()
                .flatten()
                .for_each(|expr| {
                    self.visit_expr(expr);
                }),
            syn::MacroDelimiter::Brace(_binding_0) => i
                .parse_body_with(syn::Block::parse_within)
                .ok()
                .iter()
                .flatten()
                .for_each(|stmt| {
                    self.visit_stmt(stmt);
                }),
            syn::MacroDelimiter::Bracket(_binding_0) => i
                .parse_body_with(
                    syn::punctuated::Punctuated::<syn::Expr, syn::Token![,]>::parse_terminated,
                )
                .ok()
                .iter()
                .flatten()
                .for_each(|expr| {
                    self.visit_expr(expr);
                }),
        }
    }
}
