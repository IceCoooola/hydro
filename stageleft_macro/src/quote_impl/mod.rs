use proc_macro2::{Span, TokenStream};
use quote::{quote, ToTokens};
use syn::visit::Visit;

use self::free_variable::FreeVariableVisitor;

mod free_variable;

pub fn q_impl(root: TokenStream, expr: syn::Expr) -> TokenStream {
    let mut visitor = FreeVariableVisitor::default();
    visitor.visit_expr(&expr);

    let unitialized_free_variables = visitor.free_variables.iter().map(|i| {
        let mut i_without_span = i.clone();
        i_without_span.set_span(Span::call_site());
        let ident_str = i.to_string();
        quote!(
            #[allow(unused, non_upper_case_globals, non_snake_case)]
            let #i = {
                let _out = ::#root::runtime_support::FreeVariable::uninitialized(&#i_without_span);
                _vec_to_set.push((#ident_str.to_string(), ::#root::runtime_support::FreeVariable::to_tokens(#i_without_span)));
                _out
            };
        )
    });

    let uninit_forgets = visitor.free_variables.iter().map(|i| {
        let mut i_without_span = i.clone();
        i_without_span.set_span(Span::call_site());
        quote!(
            #[allow(unused, non_upper_case_globals, non_snake_case)]
            ::std::mem::forget(#i_without_span);
        )
    });

    // necessary to ensure proper hover in Rust Analyzer
    let expr_without_spans: syn::Expr =
        syn::parse_str(&expr.clone().into_token_stream().to_string()).unwrap();

    quote!({
        move |set_mod: &mut String, set_crate_name: &mut &'static str, set_tokens: &mut #root::internal::TokenStream, _vec_to_set: &mut #root::internal::CaptureVec, run: bool| {
            #(#unitialized_free_variables;)*

            *set_mod = module_path!().to_string();
            *set_crate_name = env!("STAGELEFT_FINAL_CRATE_NAME");
            *set_tokens = #root::internal::quote! {
                #expr_without_spans
            };

            if !run {
                #(#uninit_forgets;)*
                unsafe {
                    return ::std::mem::MaybeUninit::uninit().assume_init();
                }
            }

            #[allow(unreachable_code, unused_qualifications)]
            {
                #expr
            }
        }
    })
}

#[cfg(test)]
mod tests {
    use quote::quote;
    use syn::parse_quote;

    use super::q_impl;

    macro_rules! test_quote {
        ($program:expr) => {
            let quoted_tokens = q_impl(quote!(stageleft), parse_quote!($program));
            let wrapped: syn::File = parse_quote! {
                fn main() {
                    #quoted_tokens
                }
            };

            insta::with_settings!({snapshot_suffix => "macro_tokens"}, {
                insta::assert_display_snapshot!(
                    prettyplease::unparse(&wrapped)
                );
            });
        };
    }

    #[test]
    fn test_simple() {
        test_quote! {
            1 + 2
        }
    }

    #[test]
    fn test_capture_local() {
        test_quote! {
            x + 2
        }
    }

    #[test]
    fn test_capture_copy_local() {
        test_quote! {
            (x + 2) + (x + 2)
        }
    }

    #[test]
    fn test_capture_copy_local_block() {
        test_quote! {{
            let _ = x + 2;
            let _ = x + 2;
        }}
    }

    #[test]
    fn test_capture_copy_local_block_let() {
        test_quote! {{
            let x = x + 2;
            let _ = x + 2;
        }}
    }

    #[test]
    fn test_capture_local_mut() {
        test_quote! {
            x += 2
        }
    }

    #[test]
    fn test_non_capture_local() {
        test_quote! {
            {
                let x = 1;
                x + 2
            }
        }
    }

    #[test]
    fn test_capture_in_macro() {
        test_quote! {
            dbg!(x)
        }
    }

    #[test]
    fn test_non_capture_struct_creation() {
        test_quote! {
            Foo { x: 1 }
        }
    }

    #[test]
    fn test_non_capture_enum_creation() {
        test_quote! {
            Foo::Bar { x: 1 }
        }
    }

    #[test]
    fn test_prelude_enum_variants() {
        test_quote! {
            Some(1)
        }

        test_quote! {
            None
        }

        test_quote! {
            Ok(1)
        }

        test_quote! {
            Err(1)
        }
    }
}
