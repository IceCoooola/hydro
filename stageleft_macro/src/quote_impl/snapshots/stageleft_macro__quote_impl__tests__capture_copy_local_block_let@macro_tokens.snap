---
source: stageleft_macro/src/quote_impl/mod.rs
expression: "prettyplease::unparse(&wrapped)"
---
fn main() {
    {
        move |
            set_mod: &mut String,
            set_crate_name: &mut &'static str,
            set_tokens: &mut stageleft::internal::TokenStream,
            _vec_to_set: &mut stageleft::internal::CaptureVec,
            run: bool|
        {
            #[allow(unused, non_upper_case_globals, non_snake_case)]
            let x = {
                let _out = ::stageleft::runtime_support::FreeVariable::uninitialized(&x);
                _vec_to_set
                    .push((
                        "x".to_string(),
                        ::stageleft::runtime_support::FreeVariable::to_tokens(x),
                    ));
                _out
            };
            *set_mod = module_path!().to_string();
            *set_crate_name = env!("STAGELEFT_FINAL_CRATE_NAME");
            *set_tokens = stageleft::internal::quote! {
                { let x = x + 2; let _ = x + 2; }
            };
            if !run {
                #[allow(unused, non_upper_case_globals, non_snake_case)]
                ::std::mem::forget(x);
                unsafe {
                    return ::std::mem::MaybeUninit::uninit().assume_init();
                }
            }
            #[allow(unreachable_code, unused_qualifications)]
            {
                {
                    let x = x + 2;
                    let _ = x + 2;
                }
            }
        }
    }
}
